File: typescript/13-hitl/mini-checkpoint-sim.ts
/**
 * Mini-proyecto: El checkpoint simulator (TypeScript).
 *
 * Uso:
 *   npx ts-node mini-checkpoint-sim.ts
 *   npx ts-node mini-checkpoint-sim.ts --auto
 *   npx ts-node mini-checkpoint-sim.ts --escenario destructivo --auto
 */

import * as readline from "readline";

// ── tipos ──────────────────────────────────────────────────────────────────────

// Cómo ejecutar: make ts SCRIPT=typescript/13-hitl/mini-checkpoint-sim.ts


type RiesgoNivel = "bajo" | "medio" | "alto" | "crítico";
type Decision = "pendiente" | "aprobado" | "rechazado" | "modificado";

interface Accion {
  nombre: string;
  descripcion: string;
  riesgo: RiesgoNivel;
  reversible: boolean;
  requiereHITL: boolean;
  payload: Record<string, unknown>;
}

interface Checkpoint {
  accion: Accion;
  contexto: string;
  decision: Decision;
  modificacion: Record<string, unknown>;
  latenciaS: number;
}

interface EstadoAgente {
  tarea: string;
  historial: string[];
  checkpoints: Checkpoint[];
  completado: boolean;
  abortado: boolean;
}

// ── escenarios ────────────────────────────────────────────────────────────────

const ACCIONES_SUAVES: Accion[] = [
  { nombre: "leer_archivos", descripcion: "Leer 47 archivos de configuración", riesgo: "bajo", reversible: true, requiereHITL: false, payload: { archivos: 47 } },
  { nombre: "analizar_dependencias", descripcion: "Ejecutar análisis estático de dependencias", riesgo: "bajo", reversible: true, requiereHITL: false, payload: { herramienta: "pip-audit" } },
  { nombre: "generar_reporte", descripcion: "Escribir reporte en reports/audit_2026.md", riesgo: "medio", reversible: true, requiereHITL: true, payload: { archivo: "reports/audit_2026.md", tamaño_kb: 12 } },
  { nombre: "enviar_notificacion", descripcion: "Enviar email al equipo de seguridad (3 destinatarios)", riesgo: "medio", reversible: false, requiereHITL: true, payload: { destinatarios: ["security@empresa.com"], asunto: "Audit 2026" } },
  { nombre: "cerrar_tarea", descripcion: "Marcar tarea como completada en el tracker", riesgo: "bajo", reversible: true, requiereHITL: false, payload: { ticket: "SEC-1247" } },
];

const ACCIONES_DESTRUCTIVAS: Accion[] = [
  { nombre: "listar_usuarios", descripcion: "Obtener usuarios inactivos hace >90 días", riesgo: "bajo", reversible: true, requiereHITL: false, payload: { filtro: "last_login < 90 días" } },
  { nombre: "revocar_tokens", descripcion: "Revocar tokens de 1,247 usuarios inactivos", riesgo: "alto", reversible: false, requiereHITL: true, payload: { usuarios: 1247, tokens: "API + OAuth" } },
  { nombre: "archivar_datos", descripcion: "Archivar datos en cold storage", riesgo: "alto", reversible: false, requiereHITL: true, payload: { gb: 23.4, destino: "s3://cold-archive/users/2026/" } },
  { nombre: "eliminar_cuentas", descripcion: "Eliminar definitivamente 1,247 cuentas de la BD", riesgo: "crítico", reversible: false, requiereHITL: true, payload: { usuarios: 1247, operacion: "DELETE FROM users WHERE ..." } },
  { nombre: "purgar_logs", descripcion: "Purgar logs de acceso de usuarios eliminados", riesgo: "alto", reversible: false, requiereHITL: true, payload: { registros: 89432, tabla: "access_logs" } },
];

type Escenario = { tarea: string; acciones: Accion[]; decisiones: Record<string, Decision> };

const ESCENARIOS: Record<string, Escenario> = {
  suave: {
    tarea: "Auditoría de seguridad y notificación al equipo",
    acciones: ACCIONES_SUAVES,
    decisiones: { generar_reporte: "aprobado", enviar_notificacion: "modificado" },
  },
  destructivo: {
    tarea: "Limpieza de usuarios inactivos en producción",
    acciones: ACCIONES_DESTRUCTIVAS,
    decisiones: { revocar_tokens: "aprobado", archivar_datos: "aprobado", eliminar_cuentas: "rechazado", purgar_logs: "rechazado" },
  },
};

// ── utils ─────────────────────────────────────────────────────────────────────

const ICONOS: Record<RiesgoNivel, string> = { bajo: "🟢", medio: "🟡", alto: "🔴", "crítico": "🚨" };

function mostrarCheckpoint(cp: Checkpoint): void {
  const a = cp.accion;
  console.log(`\n${"─".repeat(60)}`);
  console.log(`  CHECKPOINT — Aprobación requerida`);
  console.log("─".repeat(60));
  console.log(`  Acción:      ${a.nombre}`);
  console.log(`  Descripción: ${a.descripcion}`);
  console.log(`  Riesgo:      ${ICONOS[a.riesgo]} ${a.riesgo.toUpperCase()}`);
  console.log(`  Reversible:  ${a.reversible ? "Sí" : "NO — irreversible"}`);
  console.log(`  Contexto:    ${cp.contexto}`);
  console.log(`  Payload:     ${JSON.stringify(a.payload)}`);
  console.log("─".repeat(60));
}

async function preguntarUsuario(prompt: string): Promise<string> {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  return new Promise((resolve) => rl.question(prompt, (ans) => { rl.close(); resolve(ans); }));
}

async function solicitarDecisionInteractiva(cp: Checkpoint): Promise<Decision> {
  mostrarCheckpoint(cp);
  console.log("\n  Opciones: [A] Aprobar   [R] Rechazar   [M] Modificar   [S] Escalar");
  while (true) {
    const resp = (await preguntarUsuario("\n  Tu decisión > ")).trim().toUpperCase();
    if (resp === "A") return "aprobado";
    if (resp === "R") return "rechazado";
    if (resp === "M") { console.log("  (Modificado → dry_run: true)"); return "modificado"; }
    if (resp === "S") { console.log("  [Escalado → aprobado con flag 'escalado']"); return "aprobado"; }
    console.log("  Opción no válida.");
  }
}

function solicitarDecisionAuto(cp: Checkpoint, decisiones: Record<string, Decision>): Decision {
  mostrarCheckpoint(cp);
  const d = decisiones[cp.accion.nombre] ?? "aprobado";
  console.log(`\n  [auto] Decisión automática: ${d.toUpperCase()}`);
  return d;
}

// ── simulación ────────────────────────────────────────────────────────────────

async function simularAgente(escenario: Escenario, auto: boolean): Promise<EstadoAgente> {
  const estado: EstadoAgente = {
    tarea: escenario.tarea,
    historial: [],
    checkpoints: [],
    completado: false,
    abortado: false,
  };

  const log = (msg: string) => { estado.historial.push(msg); console.log(`  [agente] ${msg}`); };
  log(`Iniciando tarea: ${escenario.tarea}`);

  for (const accion of escenario.acciones) {
    if (estado.abortado) break;
    log(`Preparando: ${accion.nombre}`);

    if (accion.requiereHITL) {
      const cp: Checkpoint = {
        accion,
        contexto: `El agente ha completado ${estado.historial.length} pasos.`,
        decision: "pendiente",
        modificacion: {},
        latenciaS: 0,
      };
      estado.checkpoints.push(cp);

      const t0 = Date.now();
      const decision = auto
        ? solicitarDecisionAuto(cp, escenario.decisiones)
        : await solicitarDecisionInteractiva(cp);
      cp.latenciaS = (Date.now() - t0) / 1000;
      cp.decision = decision;

      if (decision === "rechazado") {
        log(`✗ ${accion.nombre} RECHAZADO — abortando`);
        estado.abortado = true;
        break;
      }
      if (decision === "modificado") cp.modificacion = { dry_run: true };
    }

    const sufijo = accion.requiereHITL && estado.checkpoints.at(-1)?.decision === "modificado"
      ? " (dry-run)" : "";
    log(`✓ ${accion.nombre} ejecutado${sufijo}`);
  }

  if (!estado.abortado) { estado.completado = true; log("Tarea completada."); }
  return estado;
}

// ── reporte ───────────────────────────────────────────────────────────────────

function imprimirReporte(estado: EstadoAgente): void {
  console.log(`\n${"=".repeat(60)}`);
  console.log("  REPORTE FINAL — CHECKPOINT SIMULATOR");
  console.log("=".repeat(60));
  console.log(`\n  Tarea: ${estado.tarea}`);
  console.log(`  Estado: ${estado.completado ? "COMPLETADA" : "ABORTADA"}`);
  console.log(`  Pasos ejecutados: ${estado.historial.length}`);
  console.log(`\n  Checkpoints (${estado.checkpoints.length} total):`);
  console.log("  " + "─".repeat(56));

  let totalLatencia = 0;
  let aprobados = 0, rechazados = 0, modificados = 0;
  for (const cp of estado.checkpoints) {
    const icon = { aprobado: "✓", rechazado: "✗", modificado: "~", pendiente: "?" }[cp.decision] ?? "?";
    console.log(`  ${icon} ${cp.accion.nombre.padEnd(30)} [${cp.decision.padEnd(10)}]  ${ICONOS[cp.accion.riesgo]} ${cp.accion.riesgo}`);
    if (cp.latenciaS > 0) { console.log(`    Latencia: ${cp.latenciaS.toFixed(1)}s`); totalLatencia += cp.latenciaS; }
    if (cp.decision === "aprobado") aprobados++;
    else if (cp.decision === "rechazado") rechazados++;
    else if (cp.decision === "modificado") modificados++;
  }

  const total = estado.checkpoints.length;
  if (total > 0) {
    const tasa = ((aprobados + modificados) / total * 100).toFixed(0);
    console.log(`\n  Tasa de aprobación: ${tasa}% (${aprobados} aprobados, ${modificados} modificados, ${rechazados} rechazados)`);
    if (parseFloat(tasa) > 95) console.log("  ⚠️  Approval fatigue — revisar umbrales de riesgo");
    if (totalLatencia > 0) console.log(`  Latencia total HITL: ${totalLatencia.toFixed(1)}s`);
  }

  console.log(`\n${"=".repeat(60)}`);
  console.log("  Lecciones:");
  console.log("  • Los checkpoints bloquean — más checkpoints = mayor latencia");
  console.log("  • Un rechazo puede abortar todo el pipeline");
  console.log("  • 'Modificar' reduce abortos sin aprobar incondicionalmente");
  console.log("  • Tasa > 95% = umbrales mal calibrados");
  console.log("=".repeat(60));
}

// ── main ──────────────────────────────────────────────────────────────────────

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const auto = args.includes("--auto");
  const escenarioKey = args.includes("--escenario") ? args[args.indexOf("--escenario") + 1] : "suave";
  const escenario = ESCENARIOS[escenarioKey] ?? ESCENARIOS.suave;

  console.log(`\n${"=".repeat(60)}`);
  console.log("  CHECKPOINT SIMULATOR");
  console.log(`  Escenario: ${escenarioKey}  |  Modo: ${auto ? "automático" : "interactivo"}`);
  console.log("=".repeat(60));
  console.log(`\n  Tarea: ${escenario.tarea}`);
  console.log(`  Checkpoints HITL: ${escenario.acciones.filter((a) => a.requiereHITL).length}`);

  const estado = await simularAgente(escenario, auto);
  imprimirReporte(estado);
}

main().catch(console.error);

File: go/13-hitl/mini-checkpoint-sim.go
// Mini-proyecto: El checkpoint simulator (Go).
//
// Uso:
//
//	go run mini-checkpoint-sim.go
//	go run mini-checkpoint-sim.go -auto
//	go run mini-checkpoint-sim.go -escenario destructivo -auto

// Cómo ejecutar: make go FILE=go/13-hitl/mini-checkpoint-sim.go

package main

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"strings"
	"time"
)

// ── tipos ─────────────────────────────────────────────────────────────────────

type RiesgoNivel string

const (
	Bajo    RiesgoNivel = "bajo"
	Medio   RiesgoNivel = "medio"
	Alto    RiesgoNivel = "alto"
	Critico RiesgoNivel = "crítico"
)

type Decision string

const (
	Pendiente  Decision = "pendiente"
	Aprobado   Decision = "aprobado"
	Rechazado  Decision = "rechazado"
	Modificado Decision = "modificado"
)

type Accion struct {
	nombre      string
	descripcion string
	riesgo      RiesgoNivel
	reversible  bool
	requiereHITL bool
	payload     map[string]interface{}
}

type Checkpoint struct {
	accion      Accion
	contexto    string
	decision    Decision
	modificacion map[string]interface{}
	latenciaS   float64
}

type EstadoAgente struct {
	tarea       string
	historial   []string
	checkpoints []Checkpoint
	completado  bool
	abortado    bool
}

func (e *EstadoAgente) log(msg string) {
	e.historial = append(e.historial, msg)
	fmt.Printf("  [agente] %s\n", msg)
}

// ── escenarios ────────────────────────────────────────────────────────────────

var accionesSuaves = []Accion{
	{nombre: "leer_archivos", descripcion: "Leer 47 archivos de configuración", riesgo: Bajo, reversible: true, requiereHITL: false, payload: map[string]interface{}{"archivos": 47}},
	{nombre: "analizar_dependencias", descripcion: "Ejecutar análisis estático de dependencias", riesgo: Bajo, reversible: true, requiereHITL: false, payload: map[string]interface{}{"herramienta": "pip-audit"}},
	{nombre: "generar_reporte", descripcion: "Escribir reporte en reports/audit_2026.md", riesgo: Medio, reversible: true, requiereHITL: true, payload: map[string]interface{}{"archivo": "reports/audit_2026.md", "tamaño_kb": 12}},
	{nombre: "enviar_notificacion", descripcion: "Enviar email al equipo de seguridad (3 destinatarios)", riesgo: Medio, reversible: false, requiereHITL: true, payload: map[string]interface{}{"destinatarios": []string{"security@empresa.com"}, "asunto": "Audit 2026"}},
	{nombre: "cerrar_tarea", descripcion: "Marcar tarea como completada en el tracker", riesgo: Bajo, reversible: true, requiereHITL: false, payload: map[string]interface{}{"ticket": "SEC-1247"}},
}

var accionesDestructivas = []Accion{
	{nombre: "listar_usuarios", descripcion: "Obtener usuarios inactivos hace >90 días", riesgo: Bajo, reversible: true, requiereHITL: false, payload: map[string]interface{}{"filtro": "last_login < 90 días"}},
	{nombre: "revocar_tokens", descripcion: "Revocar tokens de 1,247 usuarios inactivos", riesgo: Alto, reversible: false, requiereHITL: true, payload: map[string]interface{}{"usuarios": 1247, "tokens": "API + OAuth"}},
	{nombre: "archivar_datos", descripcion: "Archivar datos en cold storage", riesgo: Alto, reversible: false, requiereHITL: true, payload: map[string]interface{}{"gb": 23.4, "destino": "s3://cold-archive/users/2026/"}},
	{nombre: "eliminar_cuentas", descripcion: "Eliminar definitivamente 1,247 cuentas de la BD", riesgo: Critico, reversible: false, requiereHITL: true, payload: map[string]interface{}{"usuarios": 1247, "operacion": "DELETE FROM users WHERE ..."}},
	{nombre: "purgar_logs", descripcion: "Purgar logs de usuarios eliminados", riesgo: Alto, reversible: false, requiereHITL: true, payload: map[string]interface{}{"registros": 89432, "tabla": "access_logs"}},
}

type Escenario struct {
	tarea     string
	acciones  []Accion
	decisiones map[string]Decision
}

var escenarios = map[string]Escenario{
	"suave": {
		tarea:    "Auditoría de seguridad y notificación al equipo",
		acciones: accionesSuaves,
		decisiones: map[string]Decision{
			"generar_reporte":    Aprobado,
			"enviar_notificacion": Modificado,
		},
	},
	"destructivo": {
		tarea:    "Limpieza de usuarios inactivos en producción",
		acciones: accionesDestructivas,
		decisiones: map[string]Decision{
			"revocar_tokens":  Aprobado,
			"archivar_datos":  Aprobado,
			"eliminar_cuentas": Rechazado,
			"purgar_logs":     Rechazado,
		},
	},
}

var iconosRiesgo = map[RiesgoNivel]string{
	Bajo: "🟢", Medio: "🟡", Alto: "🔴", Critico: "🚨",
}

// ── helpers ───────────────────────────────────────────────────────────────────

func mostrarCheckpoint(cp Checkpoint) {
	a := cp.accion
	payloadJSON, _ := json.Marshal(a.payload)
	fmt.Printf("\n%s\n", strings.Repeat("─", 60))
	fmt.Println("  CHECKPOINT — Aprobación requerida")
	fmt.Println(strings.Repeat("─", 60))
	fmt.Printf("  Acción:      %s\n", a.nombre)
	fmt.Printf("  Descripción: %s\n", a.descripcion)
	fmt.Printf("  Riesgo:      %s %s\n", iconosRiesgo[a.riesgo], strings.ToUpper(string(a.riesgo)))
	if a.reversible {
		fmt.Println("  Reversible:  Sí")
	} else {
		fmt.Println("  Reversible:  NO — irreversible")
	}
	fmt.Printf("  Contexto:    %s\n", cp.contexto)
	fmt.Printf("  Payload:     %s\n", payloadJSON)
	fmt.Println(strings.Repeat("─", 60))
}

var reader = bufio.NewReader(os.Stdin)

func leerLinea() string {
	line, _ := reader.ReadString('\n')
	return strings.TrimSpace(line)
}

func solicitarDecisionInteractiva(cp Checkpoint) Decision {
	mostrarCheckpoint(cp)
	fmt.Println("\n  Opciones: [A] Aprobar   [R] Rechazar   [M] Modificar   [S] Escalar")
	for {
		fmt.Print("\n  Tu decisión > ")
		resp := strings.ToUpper(leerLinea())
		switch resp {
		case "A":
			return Aprobado
		case "R":
			return Rechazado
		case "M":
			fmt.Println("  (Modificado → dry_run: true)")
			return Modificado
		case "S":
			fmt.Println("  [Escalado → aprobado]")
			return Aprobado
		default:
			fmt.Println("  Opción no válida.")
		}
	}
}

func solicitarDecisionAuto(cp Checkpoint, decisiones map[string]Decision) Decision {
	mostrarCheckpoint(cp)
	d, ok := decisiones[cp.accion.nombre]
	if !ok {
		d = Aprobado
	}
	fmt.Printf("\n  [auto] Decisión automática: %s\n", strings.ToUpper(string(d)))
	time.Sleep(300 * time.Millisecond)
	return d
}

// ── simulación ────────────────────────────────────────────────────────────────

func simularAgente(esc Escenario, auto bool) EstadoAgente {
	estado := EstadoAgente{tarea: esc.tarea}
	estado.log("Iniciando tarea: " + esc.tarea)

	for _, accion := range esc.acciones {
		if estado.abortado {
			break
		}
		estado.log("Preparando: " + accion.nombre)

		if accion.requiereHITL {
			cp := Checkpoint{
				accion:   accion,
				contexto: fmt.Sprintf("El agente ha completado %d pasos.", len(estado.historial)),
				decision: Pendiente,
			}

			t0 := time.Now()
			var decision Decision
			if auto {
				decision = solicitarDecisionAuto(cp, esc.decisiones)
			} else {
				decision = solicitarDecisionInteractiva(cp)
			}
			cp.latenciaS = time.Since(t0).Seconds()
			cp.decision = decision

			if decision == Rechazado {
				estado.log("✗ " + accion.nombre + " RECHAZADO — abortando")
				estado.checkpoints = append(estado.checkpoints, cp)
				estado.abortado = true
				break
			}
			if decision == Modificado {
				cp.modificacion = map[string]interface{}{"dry_run": true}
			}
			estado.checkpoints = append(estado.checkpoints, cp)
		}

		sufijo := ""
		if accion.requiereHITL && len(estado.checkpoints) > 0 {
			last := estado.checkpoints[len(estado.checkpoints)-1]
			if last.accion.nombre == accion.nombre && last.decision == Modificado {
				sufijo = " (dry-run)"
			}
		}
		estado.log(fmt.Sprintf("✓ %s ejecutado%s", accion.nombre, sufijo))
	}

	if !estado.abortado {
		estado.completado = true
		estado.log("Tarea completada.")
	}
	return estado
}

// ── reporte ───────────────────────────────────────────────────────────────────

func imprimirReporte(estado EstadoAgente) {
	fmt.Printf("\n%s\n", strings.Repeat("=", 60))
	fmt.Println("  REPORTE FINAL — CHECKPOINT SIMULATOR")
	fmt.Println(strings.Repeat("=", 60))
	fmt.Printf("\n  Tarea: %s\n", estado.tarea)
	estadoStr := "ABORTADA"
	if estado.completado {
		estadoStr = "COMPLETADA"
	}
	fmt.Printf("  Estado: %s\n", estadoStr)
	fmt.Printf("  Pasos ejecutados: %d\n", len(estado.historial))
	fmt.Printf("\n  Checkpoints (%d total):\n", len(estado.checkpoints))
	fmt.Printf("  %s\n", strings.Repeat("─", 56))

	totalLatencia := 0.0
	aprobados, rechazados, modificados := 0, 0, 0
	for _, cp := range estado.checkpoints {
		icon := map[Decision]string{Aprobado: "✓", Rechazado: "✗", Modificado: "~", Pendiente: "?"}[cp.decision]
		fmt.Printf("  %s %-30s [%-10s]  %s %s\n",
			icon, cp.accion.nombre, string(cp.decision), iconosRiesgo[cp.accion.riesgo], string(cp.accion.riesgo))
		if cp.latenciaS > 0 {
			fmt.Printf("    Latencia: %.1fs\n", cp.latenciaS)
			totalLatencia += cp.latenciaS
		}
		switch cp.decision {
		case Aprobado:
			aprobados++
		case Rechazado:
			rechazados++
		case Modificado:
			modificados++
		}
	}

	total := len(estado.checkpoints)
	if total > 0 {
		tasa := float64(aprobados+modificados) / float64(total) * 100
		fmt.Printf("\n  Tasa de aprobación: %.0f%% (%d aprobados, %d modificados, %d rechazados)\n",
			tasa, aprobados, modificados, rechazados)
		if tasa > 95 {
			fmt.Println("  ⚠️  Approval fatigue — revisar umbrales de riesgo")
		}
		if totalLatencia > 0 {
			fmt.Printf("  Latencia total HITL: %.1fs\n", totalLatencia)
		}
	}

	fmt.Printf("\n%s\n", strings.Repeat("=", 60))
	fmt.Println("  Lecciones:")
	fmt.Println("  • Los checkpoints bloquean — más checkpoints = mayor latencia")
	fmt.Println("  • Un rechazo puede abortar todo el pipeline")
	fmt.Println("  • 'Modificar' reduce abortos sin aprobar incondicionalmente")
	fmt.Println("  • Tasa > 95% = umbrales mal calibrados")
	fmt.Println(strings.Repeat("=", 60))
}

// ── main ──────────────────────────────────────────────────────────────────────

func main() {
	auto := flag.Bool("auto", false, "Modo automático con decisiones simuladas")
	escenarioKey := flag.String("escenario", "suave", "Escenario: suave|destructivo")
	flag.Parse()

	esc, ok := escenarios[*escenarioKey]
	if !ok {
		esc = escenarios["suave"]
	}

	fmt.Printf("\n%s\n", strings.Repeat("=", 60))
	fmt.Println("  CHECKPOINT SIMULATOR")
	modoStr := "interactivo"
	if *auto {
		modoStr = "automático"
	}
	fmt.Printf("  Escenario: %s  |  Modo: %s\n", *escenarioKey, modoStr)
	fmt.Println(strings.Repeat("=", 60))
	fmt.Printf("\n  Tarea: %s\n", esc.tarea)
	n := 0
	for _, a := range esc.acciones {
		if a.requiereHITL {
			n++
		}
	}
	fmt.Printf("  Checkpoints HITL: %d\n", n)

	estado := simularAgente(esc, *auto)
	imprimirReporte(estado)
}
